expression/builtin_like.go

```diff
@@ -96,15 +96,25 @@ func (c *regexpFunctionClass) getFunction(ctx sessionctx.Context, args []Express
 	bf.tp.Flen = 1
 	var sig builtinFunc
 	if types.IsBinaryStr(args[0].GetType()) || types.IsBinaryStr(args[1].GetType()) {
-		sig = &builtinRegexpBinarySig{bf}
+		sig = &builtinRegexpBinarySig{builtinRegexpSharedSig{baseBuiltinFunc: bf}}
 	} else {
-		sig = &builtinRegexpSig{bf}
+		sig = &builtinRegexpSig{builtinRegexpSharedSig{baseBuiltinFunc: bf}}
 	}
 	return sig, nil
 }
 
-type builtinRegexpBinarySig struct {
+type regexpCompiler interface {
+	compile(pat string) (*regexp.Regexp, error)
+}
+
+type builtinRegexpSharedSig struct {
 	baseBuiltinFunc
+	memoizedRegexp *regexp.Regexp
+	memoizeErr     error
+}
+
+type builtinRegexpBinarySig struct {
+	builtinRegexpSharedSig
 }
 
 func (b *builtinRegexpBinarySig) Clone() builtinFunc {
```

Comment:
Should we clone `builtinRegexpSharedSig` here?

Reply:
Thanks! Fixed.